### PR TITLE
[oneDAL] Add common alg descriptor spec

### PR DESCRIPTION
**Pull request by rlnx**
Please take a look at a common section that introduces descriptors https://rlnx.github.io/oneapi-spec/programming_model/algorithm_anatomy.html

### CODE REVIEW

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 1]
```diff
@@ -1,23 +1,129 @@
...
+   auto desc = descriptor{}
+      .set_property_name_1(value_1)
+      .set_property_name_2(value_2)
+      .set_property_name_3(value_3);
+
+
+Descriptor is allowed to have any number of template parameters, but must
+support at least two:
+
+- ``Float`` is a floating-point type that the algorithm uses for computations.
```
Comment (andreynv) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:84]: This paragraph seems to deserve a separate dedicated section that would be discuss data types, etc and this default_float_t living in the library level namespace

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 2]
```diff
@@ -1,23 +1,129 @@
...
+  Algorithms are required to support only ``float`` and ``double`` types. The
+  other floating-point types, e.g., ``float16`` or ``bfloat16``, are optional
+  and up to specific oneDAL implementation. This parameter is defined first and
+  has the :expr:`onedal::default_float_t` default value.
+
+- ``Method`` is a tag-type that specifies an implementation of algorithm.
```
Comment (andreynv) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:90]: The same about this method info - dedicated section, and the descriptor would then refer to them and only specify the order of the template parameters

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 3]
```diff
@@ -1,23 +1,129 @@
...
+  Computational scheme of the algorithm may be implemented in different ways, so
+  this parameter specifies the method to use while some operation is performed.
+
+  The supported computational methods are declared within the ``method``
+  namespace nested to the algorithm's namespace. Algorithm must support at least
```
Comment (andreynv) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:95]: must->shall

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 4]
```diff
@@ -1,23 +1,129 @@
...
+  compile-time and run-time parameters.
+
+- Object that carries state of the algorithm. In the general case, descriptor is
+  stateful object whose state changes after applying an operation.
+
+
+Each oneDAL algorithm has its own dedicated namespace, where corresponding
+descriptor is defined. Descriptor, in its turn, defines a set of template
+parameters and **properties** --- run-time parameters can be accessed by means
+of the corresponding getter and setter methods. `The following code sample
```
[the remark this answers is not in the document]
Comment (rlnx) [reply source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:30]: I'm going to add them to glossary once it's merged to master

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 5]
```diff
@@ -1,23 +1,184 @@
...
+
+  .. code-block:: cpp
+
+     auto desc = descriptor{}
+        .set_property_name_1(value_1)
+        .set_property_name_2(value_2)
+        .set_property_name_3(value_3);
+
+
+Descriptor is allowed to have any number of template parameters, but shall
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:115]:
```suggestion
A descriptor is allowed to have any number of template parameters, but should
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 6]
```diff
@@ -1,23 +1,184 @@
...
+support at least two:
+
+- ``Float`` is a `floating-point type <floating-point_>`_ that the algorithm
+  uses for computations. This parameter is defined first and has the
+  ``onedal::default_float_t`` default value.
+
+- ``Method`` is a tag-type that specifies an `computational method <methods_>`_.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:122]:
```suggestion
- ``Method`` is a tag-type that specifies the `computational method <methods_>`_.
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 7]
```diff
@@ -1,23 +1,184 @@
...
+   } // namespace onedal
+
+
+.. _methods:
+
+Computational Methods
+---------------------
+
+The supported computational methods are declared within the
+``%ALGORITHM%::method`` namespace using tag-types. Algorithm shall support at
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:153]:
```suggestion
``%ALGORITHM%::method`` namespace using tag-types. Each algorithm should support at
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 8]
```diff
@@ -1,23 +1,184 @@
...
+least one computation method and declare the ``by_default`` type alias that
+refers to the one of the computational methods as shown in the example below.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:155]:
```suggestion
least one computational method and declare the ``by_default`` type alias that
refers to one of the computational methods as shown in the example below.
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 9]
```diff
@@ -1,23 +1,184 @@
...
+
+.. _floating-point:
+
+Floating-point Types
+--------------------
+
+Algorithms are required to support at least one implementation-defined
+floating-point type. The other floating-point types, e.g., ``float``,
+``double``, ``float16`` and ``bfloat16``, are optional and up to specific oneDAL
+implementation.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:135]:
```suggestion
It is required for each algorithm to support at least one implementation-defined
floating-point type. Other floating-point types are optional, for example ``float``,
``double``, ``float16``, and ``bfloat16``. It is up to a specific oneDAL
implementation whether or not to support these types.
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 10]
```diff
@@ -1,23 +1,184 @@
...
+
+The floating-point type used as a default in descriptors is
+implementation-defined and shall be declared within the top-level namespace.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:138]:
```suggestion
implementation-defined and should be declared within the top-level namespace.
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 11]
```diff
@@ -1,23 +1,184 @@
...
+Descriptor is allowed to have any number of template parameters, but shall
+support at least two:
+
+- ``Float`` is a `floating-point type <floating-point_>`_ that the algorithm
+  uses for computations. This parameter is defined first and has the
+  ``onedal::default_float_t`` default value.
+
+- ``Method`` is a tag-type that specifies an `computational method <methods_>`_.
+  This parameter is defined second and has the ``method::by_default`` default
+  value.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:124]: I think those lines are logically belong to the part where template parameters are first introduced

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 12]
```diff
@@ -1,23 +1,184 @@
...
+- Object that carries state of the algorithm. In the general case, descriptor is
+  stateful object whose state changes after applying an operation.
+
+Each oneDAL algorithm has its own dedicated namespace, where corresponding
+descriptor is defined. Descriptor, in its turn, defines a set of template
+parameters and **properties** --- run-time parameters that can be accessed by
+means of the corresponding getter and setter methods. `The following code sample
+<descriptor-template_>`_ shows the common structure of descriptor's definition
+for an abstract algorithm. To define a particular algorithm the following
+strings shall be substituted:
```
[the remark this answers is not in the document]
Comment (outoftardis) [reply source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:61]: Not sure what to add as a definition of template parameters.

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst (hunk 3 above)
Comment (outoftardis) [reply source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst]: shall is more like will, should is better

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 13]
```diff
@@ -1,23 +1,129 @@
...
+
+- Aggregator of parameters. Descriptor provides an interface for setting up both
+  compile-time and run-time parameters.
+
+- Object that carries state of the algorithm. In the general case, descriptor is
+  stateful object whose state changes after applying an operation.
+
+
+Each oneDAL algorithm has its own dedicated namespace, where corresponding
+descriptor is defined. Descriptor, in its turn, defines a set of template
```
[the remark this answers is not in the document]
Comment (outoftardis) [reply source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:56]: bullet list works fine

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 14]
```diff
@@ -1,24 +1,183 @@
...
+  meta-parameters at either compile-time or run-time.
+
+- An object that stores the state of the algorithm. In the general case, a descriptor is
+  a stateful object whose state changes after an operation is applied.
+
+Each oneDAL algorithm has its own dedicated namespace, where the corresponding
+descriptor is defined. Descriptor, in its turn, defines a set of template
+parameters and **properties** --- run-time parameters that can be accessed by
+means of the corresponding getter and setter methods. `The following code sample
+<descriptor-template_>`_ shows the common structure of descriptor's definition
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:58]:
```suggestion
<descriptor-template_>`_ shows the common structure of a descriptor's definition
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 15]
```diff
@@ -1,24 +1,183 @@
...
+for an abstract algorithm. To define a particular algorithm the following
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:59]:
```suggestion
for an abstract algorithm. To define a particular algorithm, the following
```

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 16]
```diff
@@ -51,16 +51,31 @@ computational methods. A descriptor serves as:
...
-Each oneDAL algorithm has its own dedicated namespace, where the corresponding
-descriptor is defined. Descriptor, in its turn, defines a set of template
-parameters and **properties** --- run-time parameters that can be accessed by
-means of the corresponding getter and setter methods. `The following code sample
-<descriptor-template_>`_ shows the common structure of a descriptor's definition
-for an abstract algorithm. To define a particular algorithm, the following
-strings shall be substituted:
+Each oneDAL algorithm has its own dedicated namespace (for more details, see
+:ref:`common_namespaces`), where the corresponding descriptor is defined.
+Descriptor, in its turn, defines defines the following:
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:56]: extra `defines`

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 17]
```diff
@@ -1,24 +1,186 @@
...
+  meta-parameters at either compile-time or run-time.
+
+- An object that stores the state of the algorithm. In the general case, a descriptor is
+  a stateful object whose state changes after an operation is applied.
+
+Each oneDAL algorithm has its own dedicated namespace (for more details, see
+:ref:`common_namespaces`), where the corresponding descriptor is defined.
+Descriptor, in its turn, defines defines the following:
+
+- **Template parameter.** A descriptor is allowed to have any number of template
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:58]: Why not parameters? Below you have properties in plural

File: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst [hunk 18]
```diff
@@ -1,24 +1,186 @@
...
+
+   + ``Float`` is a `floating-point type <floating-point_>`_ that the algorithm
+     uses for computations. This parameter is defined first and has the
+     ``onedal::default_float_t`` default value.
+
+   + ``Method`` is a tag-type that specifies the `computational method
+     <methods_>`_. This parameter is defined second and has the
+     ``method::by_default`` default value.
+
+- **Properties.** A property is run-time parameter that can be accessed by means
```
Comment (outoftardis) [inline source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst:69]: is a run-time

**Pull request merged**